test(nemesis): added create index nemesis

soyacz opened the pull request:
Currently we test indexes by having some c-s profiles with MV's. Indexes are created along with tables and are populated incrementally. We miss the scenario, where an user decides to create index after some time, when table is full of data.

This commit introduces new nemesis covering this scenario. When executed, gets a table with the most partitions and creates index on some random regular or static column. Waits for creation and verify if index can be used in a query without ALLOW FILTERING option without any issue. Drops index after nemesis to not change schema permanently.

refs: https://github.com/scylladb/qa-tasks/issues/1011

## PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

vponomaryov commented:
@soyacz look at unit tests failure, it has cross-imports.

soyacz commented:
> @soyacz look at unit tests failure, it has cross-imports.

I'm still testing it, I'll update with job when confirmed fully working.

fgelcer commented:
@yarongilor , it may be relevant to something you wanted to do

yarongilor commented:
> @yarongilor , it may be relevant to something you wanted to do

@fgelcer , @soyacz , @roydahan , this PR may have quite a similar coverage to another new initiated PR of https://github.com/scylladb/scylla-cluster-tests/pull/5852

So perhaps we can consider adding some missing coverage in this PR or any other idea you think of.
The main missing coverage in this PR is taking a node down prior to creating an index, then take it up afterwards.

soyacz commented:
> > @yarongilor , it may be relevant to something you wanted to do
> 
> @fgelcer , @soyacz , @roydahan , this PR may have quite a similar coverage to another new initiated PR of #5852
> 
> So perhaps we can consider adding some missing coverage in this PR or any other idea you think of. The main missing coverage in this PR is taking a node down prior to creating an index, then take it up afterwards.

Two most notable differences between our approach is
1. you create MV, I create just index on column
2. you plan to get one node down when creating MV

@roydahan @fruch @fgelcer what do you think, should I implement it this way here?

fgelcer commented:
> > > @yarongilor , it may be relevant to something you wanted to do
> > 
> > 
> > @fgelcer , @soyacz , @roydahan , this PR may have quite a similar coverage to another new initiated PR of #5852
> > So perhaps we can consider adding some missing coverage in this PR or any other idea you think of. The main missing coverage in this PR is taking a node down prior to creating an index, then take it up afterwards.
> 
> Two most notable differences between our approach is
> 
> 1. you create MV, I create just index on column
> 2. you plan to get one node down when creating MV
> 
> @roydahan @fruch @fgelcer what do you think, should I implement it this way here?

could you please evaluate the extra effort in here? 

@yarongilor , i think you should add your own test for your task, as we should not only add and remove MV during toggling node up/down, but also add some verification on correctness of data in some extent.

soyacz commented:
> > Two most notable differences between our approach is
> > 
> > 1. you create MV, I create just index on column
> > 2. you plan to get one node down when creating MV
> > 
> > @roydahan @fruch @fgelcer what do you think, should I implement it this way here?
> 
> could you please evaluate the extra effort in here?

Adding stop/start one node is pretty easy, just matter of testing how it works.
As to replace index by MV also not that hard, just different syntax (a bit more complicated) I suppose and different way of MV creation verification.
I rather wonder **if** we want to implement this way.

yarongilor commented:
> evaluate the extra effort in here?

i actually meant having both MV + SI in this nemesis. any reason not having both?

fgelcer commented:
> > evaluate the extra effort in here?
> 
> i actually meant having both MV + SI in this nemesis. any reason not having both?

for this specific nemesis, we want to test SI, so @yarongilor , please create your own nemesis with MV only (including the verification needed)

soyacz commented:
@yarongilor @fgelcer I spoke with @fruch and we agreed to not add node operations during this nemesis. There are other cases - not only getting node down that would be worth to test. General direction is to create simple disruptions and in future we'll be adding parallel nemesis. 
E.g. one idea was to run one nemesis in a loop along with disruptive ones (eg. topology changes).

soyacz commented:
I tested it, looks working fine:
https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/lukasz/job/longevity-many-clients-4h-test/34/

I decreased verbosity of index build status verification to not flood the logs.

I'm not sure if we want to backport it - it's not strictly a feature test in 5.2/2023.1 and we might want to avoid reshuffling SisyphusNemesis in these branches jobs.

fruch commented on data_dir/nemesis.yml:
```diff
@@ -97,6 +104,7 @@
   - disruptive = False
   - kubernetes = True
   - limited = True
+  - schema_changes = True
```
@roydahan you test doesn't cover the tags/labels of nemesis, hence changes in tags doesn't force update to those files

roydahan replied:
I didn't think about that...

fruch replied:
I would add it once adding the 2nd part of running from the nemesis file

fruch commented on sdcm/utils/nemesis_utils/indexes.py:
```diff
@@ -0,0 +1,98 @@
+    # get some value from table to use it in query by index
+    result = session.execute(f'SELECT "{column}" FROM {ks}.{cf} limit 1')
+    value = list(result)[0][0]
+    match type(value).__name__:
```
Those are all the possible types ?

doesn't user profile or s-b or Gemini might create other types ?

Please consider verifying this nemesis with those as well

soyacz replied:
I tried using more types, found `datetime` needed also quoting.
Generally, number of python types is much more limited than column types and I found only these 3 to require quoting or adding some reformat.

fruch replied:
you write those 3, and there only two types here in the match/case...

soyacz replied:
seems I didn't push changes... thanks for checking

yarongilor commented on sdcm/utils/nemesis_utils/indexes.py:
```diff
@@ -0,0 +1,98 @@
+
+
+def drop_index(session, ks, index_name) -> None:
+    LOGGER.info('starting dropping index: {}.{}'.format(ks, index_name))
```
lazy formatting?

soyacz replied:
fixed

soyacz commented on sdcm/utils/nemesis_utils/indexes.py:
```diff
@@ -0,0 +1,98 @@
+
+def verify_query_by_index_works(session, ks, cf, column) -> None:
+    # get some value from table to use it in query by index
+    result = session.execute(f'SELECT "{column}" FROM {ks}.{cf} limit 1')
```
here we just get value to put to `where` clause - later query all rows.

yarongilor replied:
ok, i just have a concern it might reproduce the issue of https://github.com/scylladb/scylladb/issues/12268

soyacz replied:
adding SimpleStatement with fetch_size shall help?

roydahan replied:
Yes

fruch commented:
@yarongilor can you take a look, about the things you raised ? if they are addressed

roydahan commented:
There is a conflict + one comment from @fruch about types and one adjustment to do for fetch_size.

roydahan commented:
It's a task came from a feature for 5.2, so it needs backport to 5.2.

roydahan commented on sdcm/nemesis.py:
```diff
@@ -3990,6 +3992,26 @@ def format_error_for_sla_test_and_raise(error_events):
+                wait_for_index_to_be_built(self.target_node, ks, index_name, timeout=1200)
+                verify_query_by_index_works(session, ks, cf, column)
+            finally:
+                drop_index(session, ks, index_name)
```
I think that before dropping the index, as I wrote in @yarongilor 's PR for adding MV, we should let the test run with the index for a while.
In the other PR I suggested 1-2% of the test duration, but didn't give much thought.
Maybe:
5m < 1% < 40m
(minimum 5m, 1% for the general case, maximum 40m for the very long cases).

soyacz replied:
done

soyacz commented:
I've found some issues in `get_table_with_biggest_partitions_count` - was not excluding system_ tables and not counting properly partitions, as they were separated by token ranges (I added sum and group by to the query). Fixed. Still one more run to verify all works well: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/lukasz/job/longevity-5gb-1h-MajorCompactionMonkey-aws-test/5/

soyacz commented:
One more thing: shall I rerun `create-nemesis-pipelines` or we do it when branching?

yarongilor commented on sdcm/utils/nemesis_utils/indexes.py:
```diff
@@ -0,0 +1,101 @@
+LOGGER = logging.getLogger(__name__)
+
+
+def get_table_with_biggest_partitions_count(session) -> Tuple[str, str]:
```
It is not clear why using this method. Why limiting the tables to be selected for creating SI?
IIUC, a test can have, for example, a s-b table with large-partitions, but it will never be tested if there's a c-s table with a higher number of partitions. Why not give a chance to all user tables to be tested?

soyacz replied:
You seem to be right. I was thinking that there are small ks/cf in the cluster, but apparently there are not many. I switched it to `get_non_system_ks_cf_list` with MV exclude.

yarongilor commented on sdcm/utils/nemesis_utils/indexes.py:
```diff
@@ -0,0 +1,101 @@
+    res = session.execute('select keyspace_name, table_name,  sum(partitions_count) as partitions_count'
+                          ' from system.size_estimates GROUP BY keyspace_name, table_name')
+    rows_sorted = sorted(res, key=lambda x: x.partitions_count, reverse=True)
+    row = next(row for row in rows_sorted if not row.keyspace_name.startswith("system_"))
```
Are materialized-views filtered out here?

soyacz replied:
now yes

fruch commented:
> One more thing: shall I rerun `create-nemesis-pipelines` or we do it when branching?

Yes, master should be uptodate

soyacz commented:
> > One more thing: shall I rerun `create-nemesis-pipelines` or we do it when branching?
> 
> Yes, master should be uptodate

done

soyacz commented:
Recent update:
1. now we don't filter for biggest table, just using non-system ones. 
2. replaced few logger.info with `InfoEvent` so crucial actions are in events.
3. Created jobs files.

@fruch @yarongilor please re-review

roydahan commented on sdcm/utils/common.py:
```diff
@@ -2966,3 +2966,11 @@ def __exit__(self, exit_type, value, _traceback):
+
+
+def sleep_for_percent_of_duration(duration: int, percent: int, min_duration: int, max_duration: int):
+    """Waits the percentage of a duration in seconds, with a minimum and maximum duration"""
```
Worth to explain it like min < duration * percentage < max, it's hard to understand it like this.

soyacz replied:
done

roydahan commented on test-cases/nemesis/longevity-5gb-1h-CreateIndexNemesis.yaml:
```diff
@@ -0,0 +1,31 @@
+test_duration: 90
```
Coll that you remembered to add it!
(We will need to remember this for every nemesis we add).

soyacz replied:
I didn't remember... this is automatically generated file :sweat_smile:

fruch replied:
we should be checking it somehow, in a unittest or a new step for the CI

soyacz replied:
oh, you meant the file :) yes, somehow I remembered

fruch commented on sdcm/utils/nemesis_utils/indexes.py:
```diff
@@ -0,0 +1,88 @@
+            db_event=DatabaseLogEvent.DATABASE_ERROR,
+            line="Error applying view update"):
+        session.execute(f'DROP INDEX {ks}.{index_name}')
+        time.sleep(30)  # errors can happen only within several seconds after index drop #12977
```
nitpick:
```suggestion
        time.sleep(30)  # errors can happen only within several seconds after index drop scylladb/scylladb#12977
```

fruch reviewed (approved):
LGTM
